src/emulate.c

```diff
@@ -1125,146 +1126,202 @@ static bool rv_emulate(struct riscv_t *rv, struct rv_insn_t *ir)
-        memset(&ir, 0, sizeof(struct rv_insn_t));
-
-        /* decode the instruction */
-        if (!rv_decode(&ir, insn, &rv->insn_len)) {
-            rv_except_illegal_insn(rv, insn);
-            break;
-        }
-
-        /* execute the instruciton */
-        if (!rv_emulate(rv, &ir))
-            break;
-
-        /* increment the cycles csr */
-        rv->csr_cycle++;
+    switch (opcode) {
+    case rv_insn_jal:
+    case rv_insn_jalr:
+    case rv_insn_beq:
+    case rv_insn_bne:
+    case rv_insn_blt:
+    case rv_insn_bge:
+    case rv_insn_bltu:
+    case rv_insn_bgeu:
+    case rv_insn_ecall:
+    case rv_insn_ebreak:
+    case rv_insn_mret:
+#if RV32_HAS(EXT_C)
+    case rv_insn_cj:
+    case rv_insn_cjr:
+    case rv_insn_cjal:
+    case rv_insn_cjalr:
+    case rv_insn_cbeqz:
+    case rv_insn_cbnez:
+    case rv_insn_cebreak:
+#endif
+#if RV32_HAS(Zifencei)
+    case rv_insn_fencei:
+#endif
+        return true;
     }
+    return false;
 }
 
-riscv_user_t rv_userdata(struct riscv_t *rv)
+/* hash function is used when mapping address into the block map */
+static uint32_t hash(size_t k)
 {
-    assert(rv);
-    return rv->userdata;
+    k ^= k << 21;
+    k ^= k >> 17;
+#if (SIZE_MAX > 0xFFFFFFFF)
+    k ^= k >> 35;
+    k ^= k >> 51;
+#endif
+    return k;
 }
 
-bool rv_set_pc(struct riscv_t *rv, riscv_word_t pc)
+/* allocate a basic block */
+static struct block *block_alloc(const uint8_t bits)
 {
-    assert(rv);
-#if RV32_HAS(EXT_C)
-    if (pc & 1)
-#else
-    if (pc & 3)
-#endif
-        return false;
-
-    rv->PC = pc;
-    return true;
+    struct block *block = malloc(sizeof(struct block));
+    block->insn_capacity = 1 << bits;
+    block->insn_number = 0;
+    block->predict = NULL;
+    block->ir = malloc(block->insn_capacity * sizeof(struct rv_insn_t));
+    return block;
 }
 
-riscv_word_t rv_get_pc(struct riscv_t *rv)
+/* insert a block into block map */
+static void block_insert(struct block_map *map, struct block *block)
 {
-    assert(rv);
-    return rv->PC;
+    assert(map && block);
+    const uint32_t mask = map->block_capacity - 1;
+    uint32_t index = hash(block->pc_start);
+
+    /* insert into the block map */
+    for (;; index++) {
+        if (!map->map[index & mask]) {
+            map->map[index & mask] = block;
+            break;
+        }
+    }
+    map->size++;
 }
 
-void rv_set_reg(struct riscv_t *rv, uint32_t reg, riscv_word_t in)
+/* try to locate an already translated block in the block map */
+static struct block *block_find(struct block_map *map, const uint32_t pc)
 {
-    assert(rv);
-    if (reg < RV_NUM_REGS && reg != rv_reg_zero)
-        rv->X[reg] = in;
+    assert(map);
+    uint32_t index = hash(pc);
+    const uint32_t mask = map->block_capacity - 1;
+
+    /* find block in block map */
+    for (;; index++) {
+        struct block *block = map->map[index & mask];
+        if (!block)
+            return NULL;
+
+        if (block->pc_start == pc) {
```

Comment:
Remove curly brackets (`{` and `}`) for this statement since they are not needed.